### PR TITLE
Fix standard compliance

### PR DESCRIPTION
**Pull request by crtrott**
This fixes most everything identified in the llvm tests which is wrong with our reference implementation, and where we haven't backported fixes yet from when we contributed our implementation to LLVM. 

I haven't done mdspan yet.

### CODE REVIEW

**Comment by crtrott** [comment]
Outstanding FIXMEs: 

```
../../tests/libcxx-backports/layout_left/layout_left.static_requirements.pass.cpp:  // static_assert(std::__mdspan_detail::__is_extents_v<E>); // FIXME
../../tests/libcxx-backports/layout_right/layout_right.static_requirements.pass.cpp:  //static_assert(std::__mdspan_detail::__is_extents_v<E>); //FIXME
../../tests/libcxx-backports/layout_stride/CMakeLists.txt:#mdspan_add_test(layout_stride.extents.verify) //FIXME
../../tests/libcxx-backports/layout_stride/layout_stride.ctor.strided_mapping.pass.cpp:  //test_layout<always_convertible_layout>(); //FIXME
../../tests/libcxx-backports/layout_stride/layout_stride.static_requirements.pass.cpp:  //static_assert(std::__mdspan_detail::__is_extents_v<E>); //FIXME
../../tests/libcxx-backports/mdspan/CMakeLists.txt:#mdspan_add_test(mdspan.conversion.pass) //FIXME
../../tests/libcxx-backports/mdspan/CMakeLists.txt:#mdspan_add_test(mdspan.index_operator.pass) //FIXME
../../tests/libcxx-backports/mdspan/CMakeLists.txt:#mdspan_add_test(mdspan.swap.pass) //FIXME
../../tests/libcxx-backports/mdspan/mdspan.conversion.pass.cpp:  // FIXME: these tests trigger what appears to be a compiler bug on MINGW32 with --target=x86_64-w64-windows-gnu
../../tests/libcxx-backports/mdspan/mdspan.ctor.copy.pass.cpp:  //static_assert(noexcept(MDS(m_org)) == (noexcept(H(handle))&& noexcept(M(map))&& noexcept(A(acc)))); //FIXME
../../tests/libcxx-backports/mdspan/mdspan.ctor.default.pass.cpp:    //static_assert(noexcept(MDS()) == (noexcept(H())&& noexcept(M())&& noexcept(A()))); //FIXME
../../tests/libcxx-backports/mdspan/mdspan.ctor.default.pass.cpp:  // static_assert(test()); //FIXME
../../tests/libcxx-backports/mdspan/mdspan.ctor.dh_extents.pass.cpp:    //static_assert(!std::is_constructible_v<MDS, const H&, const typename M::extents_type&>); //FIXME
../../tests/libcxx-backports/mdspan/mdspan.ctor.dh_span.pass.cpp:  //test_mdspan_ctor<mec, ac>(handle, construct_mapping(layout, std::extents<int>()), acc); //FIXME
../../tests/libcxx-backports/mdspan/mdspan.ctor.dh_span.pass.cpp:  //test_mdspan_ctor<mec, ac>(handle, construct_mapping(layout, std::extents<unsigned, 7>()), acc); //FIXME
../../tests/libcxx-backports/mdspan/mdspan.deduction.pass.cpp:  //ASSERT_SAME_TYPE(decltype(std::mdspan(handle, 5, SizeTIntType(6))), std::mdspan<T, std::dextents<size_t, 2>>); //FIXME
../../tests/libcxx-backports/mdspan/mdspan.properties.pass.cpp:  //ASSERT_SAME_TYPE(decltype(m.size()), typename MDS::size_type); //FIXME
../../tests/libcxx-backports/mdspan/mdspan.properties.pass.cpp:  //assert(!noexcept(MDS::is_always_unique())); //FIXME
../../tests/libcxx-backports/mdspan/mdspan.properties.pass.cpp:  //assert(!noexcept(MDS::is_always_exhaustive())); //FIXME
../../tests/libcxx-backports/mdspan/mdspan.properties.pass.cpp:  //assert(!noexcept(MDS::is_always_strided())); //FIXME
../../tests/libcxx-backports/mdspan/mdspan.properties.pass.cpp:  //assert(!noexcept(m.is_unique())); //FIXME
../../tests/libcxx-backports/mdspan/mdspan.properties.pass.cpp:  //assert(!noexcept(m.is_exhaustive())); //FIXME
../../tests/libcxx-backports/mdspan/mdspan.properties.pass.cpp:  //assert(!noexcept(m.is_strided())); //FIXME
../../tests/libcxx-backports/mdspan/mdspan.properties.pass.cpp:  //static_assert(test()); //FIXME
```

File: include/experimental/__p0009_bits/extents.hpp [hunk 1]
```diff
@@ -521,10 +526,13 @@ template <class IndexType, size_t... Extents> class extents {
...
              const extents<OtherIndexType, OtherExtents...> &rhs) noexcept {
-    bool value = true;
-    for (size_type r = 0; r < m_rank; r++)
-      value &= rhs.extent(r) == lhs.extent(r);
-    return value;
+    if constexpr (rank() != extents<OtherIndexType, OtherExtents...>::rank()) {
+      return false;
+    } else {
+      for (size_type r = 0; r < m_rank; r++)
+        if(rhs.extent(r) != lhs.extent(r)) return false;
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/extents.hpp]:
```suggestion
      for (size_type r = 0; r < m_rank; r++) {
        if (rhs.extent(r) != lhs.extent(r)) {
          return false;
        }
      }
```

File: include/experimental/__p0009_bits/layout_right.hpp [hunk 2]
```diff
@@ -218,15 +220,21 @@ class layout_right::mapping {
...
     friend constexpr bool operator==(mapping const& lhs, mapping<OtherExtents> const& rhs) noexcept {
       return lhs.extents() == rhs.extents();
     }
 
     // In C++ 20 the not equal exists if equal is found
 #if !(MDSPAN_HAS_CXX_20)
-    template<class OtherExtents>
+    MDSPAN_TEMPLATE_REQUIRES(
+      class OtherExtents,
+      /* requires */ ( Extents::rank() == OtherExtents::rank())
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/layout_right.hpp]:
```suggestion
      /* requires */ (Extents::rank() == OtherExtents::rank())
```

File: include/experimental/__p0009_bits/layout_stride.hpp [hunk 3]
```diff
@@ -233,7 +233,31 @@ struct layout_stride {
 
     //--------------------------------------------------------------------------------
 
-    MDSPAN_INLINE_FUNCTION_DEFAULTED constexpr mapping() noexcept = default;
+    MDSPAN_INLINE_FUNCTION_DEFAULTED constexpr mapping() noexcept
+#if defined(_MDSPAN_USE_ATTRIBUTE_NO_UNIQUE_ADDRESS)
+      : __members{
+#else
+      : __base_t(__base_t{__member_pair_t(
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/layout_stride.hpp:247]: Do we need the repeated `__base_t` here?  It's probably harmless in C++17 as the compiler can elide it, but it makes the code harder to read.

File: include/experimental/__p0009_bits/layout_stride.hpp [hunk 4]
```diff
@@ -256,7 +280,9 @@ struct layout_stride {
 #else
       : __base_t(__base_t{__member_pair_t(
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/layout_stride.hpp]: Another instance of repeated `__base_t`.

File: include/experimental/__p0009_bits/layout_stride.hpp [hunk 5]
```diff
@@ -256,7 +280,9 @@ struct layout_stride {
...
 #endif
-          e, __strides_storage_t(__impl::fill_strides(s))
+          e, __strides_storage_t([&]<size_t ... Pos>(std::index_sequence<Pos...>) {
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/layout_stride.hpp]: Lambdas with explicit template parameters require C++20.

File: include/experimental/__p0009_bits/layout_stride.hpp [hunk 6]
```diff
@@ -410,17 +436,33 @@ struct layout_stride {
...
+          else {
+            rank_type r_largest = 0;
+            for (rank_type r = 1; r < extents_type::rank(); r++)
+              if (stride(r) > stride(r_largest))
+                r_largest = r;
+            for (rank_type r = 0; r < extents_type::rank(); r++)
+              if (extents().extent(r) == 0 && r != r_largest)
+                return false;
+            return true;
+          }
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/layout_stride.hpp]:
```suggestion
          if constexpr (extents_type::rank() == 1) {
            return stride(0) == 1;
          } else {
            rank_type r_largest = 0;
            for (rank_type r = 1; r < extents_type::rank(); r++) {
              if (stride(r) > stride(r_largest)) {
                r_largest = r;
              }
            }
            for (rank_type r = 0; r < extents_type::rank(); r++) {
              if (extents().extent(r) == 0 && r != r_largest) {
                return false;
              }
            }
            return true;
          }
```

**Review by mhoemmen — CHANGES_REQUESTED** [review]
The one lambda with explicit template parameters `[&] <size_t...> { /* ... */ }` requires C++20, but it's not protected with a macro.  Otherwise, this is great! : - )

File: include/experimental/__p0009_bits/layout_stride.hpp (hunk 4 above)
Comment (crtrott) [reply include/experimental/__p0009_bits/layout_stride.hpp]: I will look at that later, revamping layout_stride itself

File: include/experimental/__p0009_bits/layout_stride.hpp (hunk 3 above)
Comment (crtrott) [reply include/experimental/__p0009_bits/layout_stride.hpp:247]: just using the pattern from elsewhere

File: include/experimental/__p0009_bits/layout_stride.hpp (hunk 5 above)
Comment (crtrott) [reply include/experimental/__p0009_bits/layout_stride.hpp]: removed

File: include/experimental/__p0009_bits/layout_stride.hpp (hunk 6 above)
Comment (crtrott) [reply include/experimental/__p0009_bits/layout_stride.hpp]: how about I run the whole thing through clang-format eventually?

File: include/experimental/__p0009_bits/extents.hpp (hunk 1 above)
Comment (crtrott) [reply include/experimental/__p0009_bits/extents.hpp]: Proposing to run the whole thing through clang-format and starting to enforce it.

File: include/experimental/__p0009_bits/layout_right.hpp (hunk 2 above)
Comment (crtrott) [reply include/experimental/__p0009_bits/layout_right.hpp]: gonna propose to run through and enforce clang-format going foward.

File: include/experimental/__p0009_bits/layout_stride.hpp [hunk 7]
```diff
@@ -444,8 +495,11 @@ struct layout_stride {
...
       bool strides_match = true;
-      for(rank_type r = 0; r < extents_type::rank(); r++)
-        strides_match = strides_match && (x.stride(r) == y.stride(r));
+      if constexpr (extents_type::rank() > 0) {
+        using common_t = std::common_type_t<index_type, typename StridedLayoutMapping::index_type>;
+        for(rank_type r = 0; r < extents_type::rank(); r++)
+          strides_match = strides_match && (static_cast<common_t>(x.stride(r)) == static_cast<common_t>(y.stride(r)));
+      }
       return (x.extents() == y.extents()) &&
              (__impl::__OFFSET(y)== static_cast<typename StridedLayoutMapping::index_type>(0)) &&
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/layout_stride.hpp]:
```suggestion
             (__impl::__OFFSET(y) == static_cast<typename StridedLayoutMapping::index_type>(0)) &&
```

**Review by mhoemmen — APPROVED** [review]
AWESOME

File: compilation_tests/ctest_layout_convertible.cpp [hunk 8]
```diff
@@ -80,10 +80,12 @@ MDSPAN_STATIC_TEST(
   !std::is_convertible<AStridedLayout<false>::mapping<E2>, LS1>::value
 );
 
+#if !MDSPAN_HAS_CXX_14 && !MDSPAN_HAS_CXX_20
 MDSPAN_STATIC_TEST(
   std::is_constructible<LS2, AStridedLayout<true>::mapping<E1>>::value &&
-  std::is_convertible<AStridedLayout<true>::mapping<E1>, LS2>::value
+  !std::is_convertible<AStridedLayout<true>::mapping<E1>, LS2>::value
```
Comment (crtrott) [inline compilation_tests/ctest_layout_convertible.cpp:86]: Only std:: layouts are convertible, everything else is explicit in the standard.

File: include/experimental/__p0009_bits/extents.hpp [hunk 9]
```diff
@@ -251,12 +251,17 @@ struct maybe_static_array {
 
 #ifdef __cpp_lib_span
   MDSPAN_TEMPLATE_REQUIRES(class T, size_t N,
-                           /* requires */ (N == m_size_dynamic))
+                           /* requires */ (N == m_size_dynamic && N > 0))
```
Comment (crtrott) [inline include/experimental/__p0009_bits/extents.hpp:254]: needed a separate constructor to make this work warning free, and make it work in constant expressions. Same strategy as with the from std::array ctor.

File: include/experimental/__p0009_bits/extents.hpp [hunk 10]
```diff
@@ -423,9 +428,9 @@ template <class IndexType, size_t... Extents> class extents {
       class OtherIndexType, size_t N,
       /* requires */
       (
-          _MDSPAN_TRAIT(std::is_convertible, OtherIndexType, index_type) &&
+          _MDSPAN_TRAIT(std::is_convertible, const OtherIndexType&, index_type) &&
```
Comment (crtrott) [inline include/experimental/__p0009_bits/extents.hpp:431]: This is what the standard says.

Comment (crtrott) [reply include/experimental/__p0009_bits/extents.hpp:431]: Same below. The LLVM tests catch the difference.

File: include/experimental/__p0009_bits/extents.hpp [hunk 11]
```diff
@@ -573,7 +582,7 @@ using dextents = typename detail::__make_dextents<IndexType, Rank>::type;
 template <class... IndexTypes>
 extents(IndexTypes...)
     -> extents<size_t,
-               size_t((IndexTypes(), ::MDSPAN_IMPL_STANDARD_NAMESPACE::dynamic_extent))...>;
+               ((void) sizeof(IndexTypes), ::MDSPAN_IMPL_STANDARD_NAMESPACE::dynamic_extent)...>;
```
Comment (crtrott) [inline include/experimental/__p0009_bits/extents.hpp:585]: Don't require default constructibility of IndexTypes to make the deduction guide work.

File: include/experimental/__p0009_bits/layout_left.hpp [hunk 12]
```diff
@@ -155,13 +157,14 @@ class layout_left::mapping {
         * other.required_span_size() is a representable value of type index_type
         */
        #if !defined(_MDSPAN_HAS_CUDA) && !defined(_MDSPAN_HAS_HIP) && !defined(NDEBUG)
-       index_type stride = 1;
-       for(rank_type r=0; r<__extents.rank(); r++) {
-         if(stride != static_cast<index_type>(other.stride(r))) {
-           // Note this throw will lead to a terminate if triggered since this function is marked noexcept
-           throw std::runtime_error("Assigning layout_stride to layout_left with invalid strides.");
```
Comment (crtrott) [inline include/experimental/__p0009_bits/layout_left.hpp:162]: complains about throw in noexcept on various compilers (warning).

File: include/experimental/__p0009_bits/layout_left.hpp [hunk 13]
```diff
@@ -203,9 +206,9 @@ class layout_left::mapping {
     MDSPAN_INLINE_FUNCTION static constexpr bool is_always_exhaustive() noexcept { return true; }
     MDSPAN_INLINE_FUNCTION static constexpr bool is_always_strided() noexcept { return true; }
 
-    MDSPAN_INLINE_FUNCTION constexpr bool is_unique() const noexcept { return true; }
-    MDSPAN_INLINE_FUNCTION constexpr bool is_exhaustive() const noexcept { return true; }
-    MDSPAN_INLINE_FUNCTION constexpr bool is_strided() const noexcept { return true; }
+    MDSPAN_INLINE_FUNCTION static constexpr bool is_unique() noexcept { return true; }
+    MDSPAN_INLINE_FUNCTION static constexpr bool is_exhaustive() noexcept { return true; }
+    MDSPAN_INLINE_FUNCTION static constexpr bool is_strided() noexcept { return true; }
```
Comment (crtrott) [inline include/experimental/__p0009_bits/layout_left.hpp:211]: This is what the standard says.

File: include/experimental/__p0009_bits/layout_left.hpp [hunk 14]
```diff
@@ -218,15 +221,21 @@ class layout_left::mapping {
       return value;
     }
 
-    template<class OtherExtents>
+    MDSPAN_TEMPLATE_REQUIRES(
+      class OtherExtents,
+      /* requires */ ( Extents::rank() == OtherExtents::rank())
+    )
```
Comment (crtrott) [inline include/experimental/__p0009_bits/layout_left.hpp:227]: constraint in the standard.

File: include/experimental/__p0009_bits/layout_left.hpp [hunk 15]
```diff
@@ -218,15 +221,21 @@ class layout_left::mapping {
...
     }
 
     // In C++ 20 the not equal exists if equal is found
 #if !(MDSPAN_HAS_CXX_20)
-    template<class OtherExtents>
+    MDSPAN_TEMPLATE_REQUIRES(
+      class OtherExtents,
+      /* requires */ ( Extents::rank() == OtherExtents::rank())
+    )
     MDSPAN_INLINE_FUNCTION
```
Comment (crtrott) [inline include/experimental/__p0009_bits/layout_left.hpp:239]: constraint in the standard.

File: include/experimental/__p0009_bits/layout_right.hpp [hunk 16]
```diff
@@ -158,13 +158,15 @@ class layout_right::mapping {
         * other.required_span_size() is a representable value of type index_type
         */
        #if !defined(_MDSPAN_HAS_CUDA) && !defined(_MDSPAN_HAS_HIP) && !defined(NDEBUG)
-       index_type stride = 1;
-       for(rank_type r=__extents.rank(); r>0; r--) {
-         if(stride != static_cast<index_type>(other.stride(r-1))) {
-           // Note this throw will lead to a terminate if triggered since this function is marked noexcept
-           throw std::runtime_error("Assigning layout_stride to layout_right with invalid strides.");
```
Comment (crtrott) [inline include/experimental/__p0009_bits/layout_right.hpp:165]: get rid of warning and make it work in constant expression

File: include/experimental/__p0009_bits/layout_right.hpp [hunk 17]
```diff
@@ -203,9 +205,9 @@ class layout_right::mapping {
     MDSPAN_INLINE_FUNCTION static constexpr bool is_always_unique() noexcept { return true; }
     MDSPAN_INLINE_FUNCTION static constexpr bool is_always_exhaustive() noexcept { return true; }
     MDSPAN_INLINE_FUNCTION static constexpr bool is_always_strided() noexcept { return true; }
-    MDSPAN_INLINE_FUNCTION constexpr bool is_unique() const noexcept { return true; }
-    MDSPAN_INLINE_FUNCTION constexpr bool is_exhaustive() const noexcept { return true; }
-    MDSPAN_INLINE_FUNCTION constexpr bool is_strided() const noexcept { return true; }
+    MDSPAN_INLINE_FUNCTION static constexpr bool is_unique() noexcept { return true; }
+    MDSPAN_INLINE_FUNCTION static constexpr bool is_exhaustive() noexcept { return true; }
+    MDSPAN_INLINE_FUNCTION static constexpr bool is_strided() noexcept { return true; }
```
Comment (crtrott) [inline include/experimental/__p0009_bits/layout_right.hpp:210]: standard says so

File: include/experimental/__p0009_bits/layout_stride.hpp [hunk 18]
```diff
@@ -109,7 +109,7 @@ struct layout_stride {
 
     //----------------------------------------------------------------------------
 
-    using __strides_storage_t = std::array<index_type, extents_type::rank()>;
+    using __strides_storage_t = detail::possibly_empty_array<index_type, extents_type::rank()>;
```
Comment (crtrott) [inline include/experimental/__p0009_bits/layout_stride.hpp:112]: get rid of use of std::array, for GPUs etc. actually this is what the LLVM implementation does too, will make the class empty for rank-0 with clang (where std:;array actually contains a char).

File: include/experimental/__p0009_bits/layout_stride.hpp [hunk 19]
```diff
@@ -158,14 +158,16 @@ struct layout_stride {
       template <class OtherExtents>
       MDSPAN_INLINE_FUNCTION
       static constexpr bool _eq_impl(mapping const& self, mapping<OtherExtents> const& other) noexcept {
-        return    _MDSPAN_FOLD_AND((self.stride(Idxs) == other.stride(Idxs)) /* && ... */)
-               && _MDSPAN_FOLD_AND((self.extents().extent(Idxs) == other.extents().extent(Idxs)) /* || ... */);
+        using common_t = std::common_type_t<index_type, typename OtherExtents::index_type>;
+        return    _MDSPAN_FOLD_AND((static_cast<common_t>(self.stride(Idxs)) == static_cast<common_t>(other.stride(Idxs))) /* && ... */)
+               && _MDSPAN_FOLD_AND((static_cast<common_t>(self.extents().extent(Idxs)) == static_cast<common_t>(other.extents().extent(Idxs))) /* || ... */);
```
Comment (crtrott) [inline include/experimental/__p0009_bits/layout_stride.hpp:163]: these and others like it eliminate warning for different signedness.

File: include/experimental/__p0009_bits/layout_stride.hpp [hunk 20]
```diff
@@ -205,6 +207,11 @@ struct layout_stride {
       }
 #endif
 
+      MDSPAN_INLINE_FUNCTION
+      static constexpr std::array<index_type, extents_type::rank()> return_strides(const __strides_storage_t& s) {
+        return std::array<index_type, extents_type::rank()>{s[Idxs]...};
+      }
```
Comment (crtrott) [inline include/experimental/__p0009_bits/layout_stride.hpp:213]: need this becasue we don't store a std::array for strides anymore.

File: include/experimental/__p0009_bits/layout_stride.hpp [hunk 21]
```diff
@@ -233,7 +240,31 @@ struct layout_stride {
 
     //--------------------------------------------------------------------------------
 
-    MDSPAN_INLINE_FUNCTION_DEFAULTED constexpr mapping() noexcept = default;
+    MDSPAN_INLINE_FUNCTION_DEFAULTED constexpr mapping() noexcept
+#if defined(_MDSPAN_USE_ATTRIBUTE_NO_UNIQUE_ADDRESS)
+      : __members{
+#else
+      : __base_t(__base_t{__member_pair_t(
+#endif
```
Comment (crtrott) [inline include/experimental/__p0009_bits/layout_stride.hpp:248]: Need to create "layout_right" strides in default construction.

File: include/experimental/__p0009_bits/layout_stride.hpp [hunk 22]
```diff
@@ -332,10 +363,10 @@ struct layout_stride {
...
 #endif
     MDSPAN_CONDITIONAL_EXPLICIT(
-      (!std::is_convertible<typename StridedLayoutMapping::extents_type, extents_type>::value) &&
-      (detail::__is_mapping_of<layout_left, StridedLayoutMapping> ||
-       detail::__is_mapping_of<layout_right, StridedLayoutMapping> ||
-       detail::__is_mapping_of<layout_stride, StridedLayoutMapping>)
+      !(std::is_convertible<typename StridedLayoutMapping::extents_type, extents_type>::value &&
+       (detail::__is_mapping_of<layout_left, StridedLayoutMapping> ||
+        detail::__is_mapping_of<layout_right, StridedLayoutMapping> ||
+        detail::__is_mapping_of<layout_stride, StridedLayoutMapping>))
```
Comment (crtrott) [inline include/experimental/__p0009_bits/layout_stride.hpp:369]: this was just wrong.

Comment (crtrott) [reply include/experimental/__p0009_bits/layout_stride.hpp:369]: LLVM tests catch this.

File: include/experimental/__p0009_bits/layout_stride.hpp [hunk 23]
```diff
@@ -410,17 +441,37 @@ struct layout_stride {
...
+              if (extents().extent(r) == 0 && r != r_largest) {
+                return false;
+              }
+            }
+            return true;
+          }
+        } else {
+          return required_span_size() == __get_size(extents(), std::make_index_sequence<extents_type::rank()>());
+        }
+      }
```
Comment (crtrott) [inline include/experimental/__p0009_bits/layout_stride.hpp:468]: Now fully implement corner case required_span_size.

File: include/experimental/__p0009_bits/layout_stride.hpp [hunk 24]
```diff
@@ -444,10 +495,13 @@ struct layout_stride {
     MDSPAN_INLINE_FUNCTION
     friend constexpr bool operator==(const mapping& x, const StridedLayoutMapping& y) noexcept {
       bool strides_match = true;
-      for(rank_type r = 0; r < extents_type::rank(); r++)
-        strides_match = strides_match && (x.stride(r) == y.stride(r));
+      if constexpr (extents_type::rank() > 0) {
+        using common_t = std::common_type_t<index_type, typename StridedLayoutMapping::index_type>;
+        for(rank_type r = 0; r < extents_type::rank(); r++)
+          strides_match = strides_match && (static_cast<common_t>(x.stride(r)) == static_cast<common_t>(y.stride(r)));
+      }
```
Comment (crtrott) [inline include/experimental/__p0009_bits/layout_stride.hpp:502]: get rid of warning

File: include/experimental/__p0009_bits/mdspan.hpp [hunk 25]
```diff
@@ -125,8 +125,8 @@ class mdspan
   MDSPAN_TEMPLATE_REQUIRES(
     class SizeType, size_t N,
     /* requires */ (
-      _MDSPAN_TRAIT(std::is_convertible, SizeType, index_type) &&
-      _MDSPAN_TRAIT(std::is_nothrow_constructible, index_type, SizeType) &&
+      _MDSPAN_TRAIT(std::is_convertible, const SizeType&, index_type) &&
+      _MDSPAN_TRAIT(std::is_nothrow_constructible, index_type, const SizeType&) &&
```
Comment (crtrott) [inline include/experimental/__p0009_bits/mdspan.hpp:129]: this is what the standard actually says.

**Pull request merged**